0_From_Code_First_To_First_Principles/1_code_first/assignment/src/controllers/userController.ts

```diff
@@ -0,0 +1,191 @@
+        tempUser.password = user.password
+
+        const userValidationFailure = validateUser(tempUser);
+        if (userValidationFailure != null)
+        {
+            return userValidationFailure;
+        }
+
+        if (user.username != tempUser.username)
+        {
+            const usernameValidationFailure = validateUsername(tempUser.username);
+            if (await usernameValidationFailure != null)
+            {
+                return usernameValidationFailure;
+            }
+        }
+
+        if (user.email != tempUser.email)
+        {
+            const emailValidationFailure = validateEmail(tempUser.email);
+            if (await emailValidationFailure != null)
+            {
+                return emailValidationFailure;
+            }
+        }
+        
+        user.email = tempUser.email;
+        user.username = tempUser.username
+        user.firstName = tempUser.firstName;
+        user.lastName = tempUser.lastName;
+    
+        await userRepository.save(user)
+        const result = new RequestResult(200, true, undefined, getUserDataJson(user));
+        return result;
+    }
+    catch (error) {
+        console.log(error);
+        const result = new RequestResult(500, false, "ServerError", undefined);
+        return result;
+    }
+}
+
+export async function getUserByEmail(userEmail: string)
+{
+    try {
+        const userRepository = AppDataSource.getRepository(User);
+        const user = await userRepository.findOneBy({
+            email: userEmail
+        });
+        if (user == null)
+        {
+            const userNotFoundResult = getUserNotFoundResult();
+            return userNotFoundResult;
+        }
+
+        const result = new RequestResult(200, true, undefined, getUserDataJson(user));
+        return result;
+    }
+    catch (error) {
+        console.log(error);
+        const result : any = new RequestResult(500, false, "ServerError", undefined);
+        return result;
+    }
+}
+
+function randomString(n: number)
+{
+    var s = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
+    return Array(n).join().split(',').map(function() { return s.charAt(Math.floor(Math.random() * s.length)); }).join('');
+}
+
+function validateUser(user: User)
+{
+    if (!user.email || !user.username || !user.firstName || !user.lastName || !user.password)
+    {
+        const result = new RequestResult(400, false, "ValidationError", undefined);
+        return result;
+    }
+    return null;
+}
+
+async function validateUsername(username: string)
+{
+    const userRepository = AppDataSource.getRepository(User);
+    const userWithSameUsername = await userRepository.findOneBy({
+        username: username,
+    })
+    if (userWithSameUsername != null)
+    {
+        const result = new RequestResult(409, false, "UsernameAlreadyTaken", undefined);
+        return result;
+    }
+    return null;
+}
+
+async function validateEmail(email: string)
```

Comment:
You may consider using a validation library such as `zod `or `joi` for more detailed validations.